src/downsampler.ts

```diff
@@ -0,0 +1,239 @@
+    }
+    const status = pvDownsamplerInit(
+      inputFrequency,
+      outputFrequency,
+      order,
+      objectAddressAddress,
+    );
+    if (status !== 0) {
+      throw new Error(`pv_downsampler_init failed with status ${status}`);
+    }
+    const memoryBufferView = new DataView(memory.buffer);
+    const objectAddress = memoryBufferView.getInt32(objectAddressAddress, true);
+
+    const inputframeLength = pvDownsamplerConvertNumSamplesToInputSampleRate(
+      objectAddress,
+      frameLength,
+    );
+    const inputBufferAddress = alignedAlloc(
+      Int16Array.BYTES_PER_ELEMENT,
+      inputframeLength * Int16Array.BYTES_PER_ELEMENT,
+    );
+    if (inputBufferAddress === 0) {
+      throw new Error('malloc failed: Cannot allocate memory');
+    }
+    const outputBufferAddress = alignedAlloc(
+      Int16Array.BYTES_PER_ELEMENT,
+      frameLength * Int16Array.BYTES_PER_ELEMENT,
+    );
+    if (outputBufferAddress === 0) {
+      throw new Error('malloc failed: Cannot allocate memory');
+    }
+
+    const pvDownsamplerReset = instance.exports
+      .pvDownsamplerReset as CallableFunction;
+    const pvDownsamplerProcess = instance.exports
+      .pv_downsampler_process as CallableFunction;
+    const pvDownsamplerDelete = instance.exports
+      .pv_downsampler_delete as CallableFunction;
+
+    return {
+      inputBufferAddress: inputBufferAddress,
+      inputframeLength: inputframeLength,
+      memory: memory,
+      objectAddress: objectAddress,
+      outputBufferAddress: outputBufferAddress,
+      pvDownsamplerConvertNumSamplesToInputSampleRate:
+        pvDownsamplerConvertNumSamplesToInputSampleRate,
+      pvDownsamplerInit: pvDownsamplerInit,
+      pvDownsamplerProcess: pvDownsamplerProcess,
+      pvDownsamplerReset: pvDownsamplerReset,
+      pvDownsamplerDelete: pvDownsamplerDelete,
+    };
+  }
+
+  public process(
+    inputBuffer: Int16Array,
+    inputBufferSize: number,
+    outputBuffer: Int16Array,
+  ): number {
+    this._memoryBuffer.set(
+      inputBuffer,
+      this._inputBufferAddress / Int16Array.BYTES_PER_ELEMENT,
+    );
+
+    const processedSamples = this._pvDownsamplerProcess(
+      this._objectAddress,
+      this._inputBufferAddress,
+      inputBufferSize,
+      this._outputBufferAddress,
+    );
+    for (let i = 0; i < processedSamples; i++) {
+      outputBuffer[i] = this._memoryBufferView.getInt16(
+        this._outputBufferAddress + i * Int16Array.BYTES_PER_ELEMENT,
+        true,
+      );
+    }
+    return processedSamples;
+  }
+
+  public reset(): void {
+    this._pvDownsamplerReset(this._objectAddress);
+  }
+
+  public delete(): void {
+    this._pvDownsamplerDelete(this._objectAddress);
+  }
+
+  public getNumRequiredInputSamples(numSample: number): number {
+    return this._pvDownsamplerConvertNumSamplesToInputSampleRate(
+      this._objectAddress,
+      numSample,
+    );
+  }
+
+  private static arrayBufferToStringAtIndex(
```

Comment:
this method and the one below don't need to be part of the class; you can move them outside. they can live in this file, or they could be put in a `utils.ts` file and imported. Either is acceptable.